schnorr_fun/src/frost.rs

```diff
@@ -473,30 +438,50 @@ impl<H: Digest<OutputSize = U32> + Clone, NG: NonceGen> Frost<H, NG> {
 
     /// Generate nonces for creating signatures shares.
     ///
+    /// ⚠ You must use a CAREFULLY CHOSEN nonce rng, see [`FrostKey::gen_nonce_rng`]
+    pub fn gen_nonce<R: RngCore>(&self, nonce_rng: &mut R) -> NonceKeyPair {
+        NonceKeyPair::random(nonce_rng)
+    }
+
+    /// Generate a reusable nonce rng.
+    ///
     /// This method should be used carefully.
-    /// This calls [`NonceKeyPair::generate`] internally with the frost instance's [`NonceGen`].
-    /// See documentation for that for more usage info.
     ///
     /// When choosing a `secret` to use, if you are generating nonces prior to [`KeyGen`] completion,
     /// use the static first coefficient of your polynomial.
     /// Otherwise you can use your secret share of the frost key.
     ///
     /// The application should decide upon a unique `session_id`. If the `NonceGen` of this FROST
     /// instance is `Deterministic` then the `session_id` **must** be unique per signing session.
-    pub fn gen_nonce(
+    pub fn gen_nonce_rng<R: SeedableRng>(
         &self,
         secret: &Scalar,
         session_id: &[u8],
         public_key: Option<Point<impl Normalized>>,
         message: Option<Message<'_>>,
-    ) -> NonceKeyPair {
-        NonceKeyPair::generate(
-            self.schnorr.nonce_gen(),
-            secret,
-            session_id,
-            public_key,
-            message,
-        )
+    ) -> R
+    where
+        <R as SeedableRng>::Seed: From<GenericArray<u8, U32>>,
+    {
+        let message = message.unwrap_or(Message::raw(b""));
+        let msg_len = (message.len() as u64).to_be_bytes();
+        let sid_len = (session_id.len() as u64).to_be_bytes();
+        let pk_bytes = public_key
+            // NOTE: the `.normalize` here is very important. Even though the public key is already
+            // normalized we want it in particular to be Normal so that it serialzes correctly
+            // regardless of whether you pass in a Normal or EvenY point.
+            .map(|public_key| {
+                let public_key: Point<Normal> = public_key.normalize();
+                public_key.to_bytes()
+            })
+            .unwrap_or([0u8; 33]);
+        let rng: R = derive_nonce_rng!(
+            nonce_gen => self.schnorr.nonce_gen(),
+            secret => &secret,
+            public => [b"r1", pk_bytes, msg_len, message, sid_len, session_id],
```

Comment:
"r1" doesn't make sense here and potentially harms the security. We have to guarantee that it is impossible to generate the same inputs as normal single signer `Schnorr`. I think this is guaranteed if we make the `pk_bytes` the first thing in.

Actually it would be ideal if we did this properly and simply cloned the nonce gen out of the `self.schnorr` and tagged it with `frost/nonce` and stored it in `Frost`. It already has the `NonceGen` type as a parameter so no real friction here.